### DISCUSSION TITLE
Support high resolution profile images

### DISCUSSION THREAD

**Discussion by brootle** [deso-protocol/dips#174] [Medium CIPs]
Add support for high resolution profile images. 

Add API mechanism that would allow to add url for image associated with user profile.

This must be done on backend side. There is no point for 100+ apps to invent their own way to do it, once it's done, 100+ apps will be able to render nice profile images. User would update his/her profile image via any app and can be sure that he looks nice on all apps built on DeSo

Look at @brian_armstrong profile at Instagram style app that I am building, his profile images is so ugly https://gemstori.com/brian_armstrong Why actors would want to join DeSo and see their faces being brootally uglified?

![image](https://user-images.githubusercontent.com/20090030/145941512-4906b181-6ae7-4bfa-941b-d13eabc35d77.png)

**Comment by znmead** [comment]
Fully support this, we have all of the reasons to offer the incoming masses the best possible experience.

**Comment by ItsAditya-xyz** [comment]
A good solution would be to have an option for ProfileURL while still keeping the base64 images on chain. just like we have a pointer for arweave image URLs.

**Comment by tijno** [comment]
2 possible additional solutions to @AdityaChaudhary0005 :

* Link profile image to owned NFT - should not be too hard to implement & would be cool if DESO delivered this before Twitter. If user no longer owns the NFT then it reverts to the on chain image.

* Use gravatar for users that confirmed their email? 

## nft

all it needs i think is the following:

* option in NFT drop down menu to "set as profile picture"
* additional field in UPDATE_PROFILE transaction to set nft post hash
* check in backend to ensure user owns NFT and returning the stored image of the NFT as pfp

## gravatar

all it needs is a md5 of the lowercase email.

the backend could return this with profiles without exposing the email address

![profile](https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50?s=200)

`https://www.gravatar.com/avatar/205e460b479e2e5b48aec07710c08d50?s=200`

**Reply by stet** [reply]
yeah that sums up what i mentioned in discord, thanks for posting it cleanly.

**Comment by HPaulson** [comment]
This is a solid idea, one that everyone's thought about, but tread lightly; We have to find the best of both worlds here. 

I like @tijno's idea re: Gravatar & NFT images, but definitely think this one needs extensive conversation to weigh benefits

**Reply by tijno** [reply]
what could be the downsides:

gravatar 

- someone reversing md5?

- users use gravatar for more professional pfp on like work emails and they are not aware it shows in deso (so would need to be disabled by default)

- forcing people to reg their email just to get better pfp

nft

- size of image

cant think of any other negatives

**Reply by HPaulson** [reply]
I think the main negative is the same reason why PFP's are stored on-chain now. Profile images are a vital aspect of someone's social identity and having them exist in a centralized off-chain space probably isn't the best move. I'll also note that with Gravatar, they'd have to verify their email on *every* node, not just one, which is a hassle and a half.

**Comment by brootle** [comment]
I already proposed a solution, users must have ability to save url to image file as their profile image, it can be NFT, it can be gravatar, it can be hosted at his home computer, it does not matter. I own this profile, I want to have full control and be able to put any url in "Profile image field" or whatever it can be called. Image can be stored at arweave decentralized storage. I don't get it why everybody so worried that actual file will be stored somewhere outside DeSo, I just want my profile to look nice and you don't seem to worry that 99% of DeSo NFTs have images as urls to centralize image storage.

@maebeam @diamondhands0 I am ready to help you to architect better social part of DeSo protocol, send me tickets to LA, thanks!

**Reply by stet** [reply]
yeah an arbitrary url is also proper but then I might suggest letting any post allow for ANY url to image, then the update profile transaction can still be based on post hash. on Twitter when you update your profile pic it suggests you post a tweet with it. This is the opposite flow, but it shows that a correlation with a profile pic and post is a familiar pattern. 

but i get your point... why not just allow for any url without any post association. it would be fine, but then we would want both options... esp because verified NFT PFP is becoming a core feature on all platforms sooner or later. So the question is then, why would we need both? Is posting your new profile pic abnormal? I dont think so.  if we are going to implement a change to point to higher quality images for profile pics, then lets cover all bases, not just partial solution.

**Reply by HPaulson** [reply]
@brootle Using *any* URL for the images would force frontends to request literally any possible image URL someone might use as a profile image, and they must include such in their allowed img-src. This is INCREDIBLY dangerous on an extensive number of levels and is a recipe for disaster. Unless you're a fan of XSS attacks, this is a horrible horrible idea. THIS is the exact reason why we have to be very careful when making changes like these -- especially radical ones like you provided above. 

Now, you could of course argue that this is the same as post images - but it's not; Failure to display one or two posts from a "trusted" image source is a minor inconvenience, but not displaying profile images is a *huge* inconvenience. This definitely isn't the best solution.

**Reply by brootle** [reply]
why is this my problem if some idiot pushed invalid URL as his profile image? even bitclout and diamondapp have fallback url for profile images and display some image in case profile image is not available, and no problem with XSS, anything can be injected in post body, front end apps must sanitize content if they are inserting it as html

**Comment by mubashariqbal** [comment]
I like the purpose here, but I think a more general solution would be to add UserExtraData, similar to PostExtraData.

This would allow nodes to store additional data they can use for their needs.  

Need a high res image, add a url to the UserExtraData. 

Need a profile header image, add a url to the UserExtraData. 

Need a pinned post? Add a post hash to the UserExtraData.

Etc.

**Reply by tijno** [reply]
Me like this idea.

Pinned user posts pretty sure were virtually done in one of the branches.

**Reply by brootle** [reply]
I am surprised that there is no ExtraData inside Profile, that would be good to add... also see my post below, ImageURLs field added to Profile too, so we have standard across all DeSo ecosystem that devs can rely on, otherwise each app dev will invent their own way how to store high res avatar in ExtraData  and if user would add photo with one app, it will not work with another app because another app decided that this object that holds url to profile image should have different set of fields and values

**Reply by stet** [reply]
yes, this would also be ideal for most flexibility, but as stated before, there MUST be an option as part of this that references an NFT PFP because the ecosystem is moving towards this. So the standard should include a difference between any image url and an NFT post image url. the UI should have an icon displayed too when it is a verified owned NFT that is used for profile pic.

**Reply by brootle** [reply]
I am totally fine if give a choice to a user to 
a. Upload totally new image and save it in DeSo image storage
b. Let the user select existing image based on his previous posts
c. Let the user insert url for image that can be stored anywhere

**Reply by brootle** [reply]
@superzordon can you please fix this for DiamondApp? thanks!

**Reply by HPaulson** [reply]
I second this idea, but ONLY under the condition noted by @brootle above. The extra data field *needs* to be standardized. This will be useful for a whole ton of features, but only if we have spec of fields across apps.

For images specifically, I still don't think using *any* URL is the best solution (See my comment in the above thread). Would love more opinions on this topic

**Reply by alanzucker** [reply]
structured data schema for profiles ... https://schema.org/ProfilePage

**Comment by brootle** [comment]
I think I should just demonstrate what I am talking about. Just add **ImageURLs** inside **Profile** object for _get-single-profile_ API call response
![image](https://user-images.githubusercontent.com/20090030/146017993-186973df-c49c-4f71-a7a4-c8812084a9c4.png)

**Reply by DegenDudes** [reply]
This is a no brainer. Imagine if IG used low resolution on profile images 🤣

But furthermore, follow the decentralized standard and stop trying to hand away control of our data with inherited centralized coding methods.

**Comment by brootle** [comment]
@diamondhands0 @maebeam let's approve and move it into todo list 😉

**Comment by diamondhands0** [comment]
Hey guys, thanks for this prompting. I wanted to let everyone know that, thanks to this discussion, we are prioritizing a ProfileExtraData field that should achieve this, as well as allow node operators to add any custom fields they want to user profiles.

I'm sorry it takes a while to get our attention sometimes, but this seems like a clear win and we're on it.

**Reply by tijno** [reply]
would it be an idea to have some standard for extra data - its a free for all on posts right now.

https://bitclout.com/posts/02f48d617774d6d6115720b682d32b7ed4bb3efb162a74aa387432cdb96c50b1

Also https://discord.com/channels/914262395315179540/919180713746645062

I have done some checks on postgres instance (synced till accidental hardfork)

Unique key data on post extra data:
401 keys used < 500 times
358 keys used < 100 times
251 keys used <  10 times.
153 keys used  1 time

here are the top keys used:

<img width="187" alt="CleanShot 2022-01-25 at 08 03 10@2x" src="https://user-images.githubusercontent.com/69529928/150936114-5272e86b-9c7b-4638-bd61-4b33540c7808.png">

**Reply by brootle** [reply]
@diamondhands0 please, read again my comment https://github.com/deso-protocol/dips/discussions/174#discussioncomment-1805728
simply add **_ImageURLs_** inside **_Profile_** object 

ProfileExtraData will not solve the problem, it will create more problems! There must be unified way of storing high res user profile image. If you plan to let apps save JSON into ProfileExtraData it means 200 apps will be saving what they think works best for them. So at the end of the day there will be 200 different ways for storing user profile image

**Reply by chafreaky** [reply]
Can't believe I am saying this but... I agree with the above. ha

We need somewhat of a standard, otherwise things will get ugly, fast. They already are starting to be.

**Reply by darian-parrish** [reply]
I agree with the notion of a standard but here a standard convention might be better than a forced standard so flexibility is available when needed.

**Reply by dharmeshs** [reply]
Sorry I'm late to the party, but I really think this should be a *standard* URL field on the user profile (and thereby a column in the pg_profiles table).  

I like the extra user profile data too, but that'll be useful for other things and add a lot of power/flexibility.

What we need *here* (for profile images) is a standard field so that all clients/nodes can expect it to be there and handle it the same way.

Solve for simplicity and standardization.

**Comment by oO** [comment]
One simple solution could be:

1. User submits a higher resolution image to the backend to update their profile
2. Image gets stored on images.deso.org as `https://images.deso.org/u/<pkey>`
3. 100x100 version gets burned into the block chain, as of now.

`GET .../api/v0/get-single-profile-picture/<pkey>` returns the current 100x100 version
`GET .../api/v0/get-single-profile-picture/<pkey>/original` returns the original, off-chain version. ( by providing a URL redirect? )

No change to the UserProfile, no need to rely on ExtraData. ( decentralization of images.deso.org being a separate problem to solve )
